go/src/lib.rs

```diff
@@ -1,9 +1,159 @@
+#![deny(unsafe_op_in_unsafe_fn)]
+
 /**
  * Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
  */
-use std::ffi::c_void;
+use glide_core::client::Client as GlideClient;
+use glide_core::connection_request;
+use glide_core::errors;
+use glide_core::errors::RequestErrorType;
+use protobuf::Message;
+use std::{
+    ffi::{c_void, CString},
+    os::raw::c_char,
+};
+use tokio::runtime::Builder;
+use tokio::runtime::Runtime;
+
+/// Success callback that is called when a Redis command succeeds.
+// TODO: Change message type when implementing command logic
+pub type SuccessCallback =
+    unsafe extern "C" fn(channel_address: usize, message: *const c_char) -> ();
+
+/// Failure callback that is called when a Redis command fails.
+///
+/// `error` should be manually freed after this callback is invoked, otherwise a memory leak will occur.
+pub type FailureCallback = unsafe extern "C" fn(
+    channel_address: usize,
+    error_message: *const c_char,
+    error_type: RequestErrorType,
+) -> ();
+
+/// The connection response.
+///
+/// It contains either a connection or an error. It is represented as a struct instead of an enum for ease of use in the wrapper language.
+///
+/// This struct should be freed using `free_connection_response` to avoid memory leaks.
+#[repr(C)]
+pub struct ConnectionResponse {
+    conn_ptr: *const c_void,
+    error_message: *const c_char,
+    error_type: RequestErrorType,
+}
+
+/// The glide client.
+// TODO: Remove allow(dead_code) once command logic is implemented
+#[allow(dead_code)]
+pub struct Client {
+    client: GlideClient,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+    runtime: Runtime,
+}
+
+struct CreateClientError {
+    message: String,
+    error_type: RequestErrorType,
+}
+
+fn create_client_internal(
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, CreateClientError> {
+    let request = connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes)
+        .map_err(|err| CreateClientError {
+            message: err.to_string(),
+            error_type: RequestErrorType::Unspecified,
+        })?;
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_name("GLIDE for Redis Go thread")
+        .build()
+        .map_err(|err| {
+            let redis_error = err.into();
+            CreateClientError {
+                message: errors::error_message(&redis_error),
+                error_type: errors::error_type(&redis_error),
+            }
+        })?;
+    let _runtime_handle = runtime.enter();
+    let client = runtime
+        .block_on(GlideClient::new(request))
+        .map_err(|err| CreateClientError {
+            message: err.to_string(),
+            error_type: RequestErrorType::Disconnect,
```

Comment:
Similar to the above comment, this normally would return ConnectionError, but in this case, since there is no variant for that in the `RequestErrorType` enum, we went with Disconnect. Is it worth adding a ConnectionError variant to the core?